neo/Consensus/RecoveryMessage.cs

```diff
@@ -0,0 +1,104 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using Neo.IO;
+using Neo.Network.P2P.Payloads;
+
+namespace Neo.Consensus
+{
+    internal class RecoveryMessage : ConsensusMessage
+    {
+        public byte[][] ChangeViewWitnessInvocationScripts;
+        public uint[] ChangeViewTimestamps;
+        public byte[] OriginalChangeViewNumbers;
+
+        // The following 4 fields are to be able to regenerate the PrepareRequest.
+        public UInt256[] TransactionHashes;
+        // The following 3 fields are not serialized if TransactionHashes is null, indicating there is
+        // no PrepareRequest present in the RecoveryMessage.
+        public ulong Nonce;
+        public UInt160 NextConsensus;
+        public MinerTransaction MinerTransaction;
+
+        /// The PreparationHash in case the PrepareRequest hasn't been received yet.
+        /// This can be null if the PrepareRequest information is present, since it can be derived in that case.
+        public UInt256 PreparationHash;
+        public byte[][] PrepareWitnessInvocationScripts;
+        public uint[] PrepareTimestamps;
+        public byte[][] CommitSignatures;
+
+        public RecoveryMessage() : base(ConsensusMessageType.RecoveryMessage)
+        {
+        }
+
+        public override void Deserialize(BinaryReader reader)
+        {
+            base.Deserialize(reader);
+            ChangeViewWitnessInvocationScripts = reader.ReadVarBytesArray();
+            if (ChangeViewWitnessInvocationScripts != null)
```

Comment:
Is it possible for `ChangeViewWitnessInvocationScripts` to be `null`?

Reply:
Yes

Reply:
If it is sending a recovery where `M` preparations are present there is no need to send any information about the `ChangeView` messages any more, so it saves space by not including them.